File: src/runtime/javascript/engine/workflow/graph.ts
import { Table } from "../table";
import type { Workflow, WorkflowEdge, WorkflowNode } from "./types";

const NOTEBOOK_VISUAL_EDGE_ROLES = new Set(["notebook-order", "notebook-variable", "notebook-parameter", "notebook-provenance"]);

export function isDataEdge(edge: WorkflowEdge): boolean {
  return !NOTEBOOK_VISUAL_EDGE_ROLES.has(String(edge.data?.role ?? ""));
}

export function dataEdges(workflow: Workflow): WorkflowEdge[] {
  return workflow.edges.filter(isDataEdge);
}

export function orderedNodes(workflow: Workflow): WorkflowNode[] {
  const nodes = workflow.nodes;
  const edges = dataEdges(workflow);
  if (nodes.length && nodes.every((node) => typeof node.data.parameters.notebookCellIndex === "number" || node.data.parameters.notebookCellIndex !== undefined)) {
    const withIndex = nodes.map((node) => ({
      node,
      cell: Number(node.data.parameters.notebookCellIndex ?? 0),
      operation: Number(node.data.parameters.notebookOperationIndex ?? 0),
    }));
    withIndex.sort((a, b) => a.cell - b.cell || a.operation - b.operation);
    return withIndex.map((item) => item.node);
  }
  const byId = new Map(nodes.map((node) => [node.id, node]));
  if (byId.size !== nodes.length) throw new Error("Workflow node IDs must be unique");
  const indegree = new Map(nodes.map((node) => [node.id, 0]));
  const downstream = new Map<string, string[]>();
  for (const edge of edges) {
    if (!byId.has(edge.source) || !byId.has(edge.target)) throw new Error("Workflow contains an edge with a missing node");
    const list = downstream.get(edge.source) ?? [];
    list.push(edge.target);
    downstream.set(edge.source, list);
    indegree.set(edge.target, (indegree.get(edge.target) ?? 0) + 1);
  }
  const queue = nodes.filter((node) => indegree.get(node.id) === 0).map((node) => node.id);
  const ordered: WorkflowNode[] = [];
  while (queue.length) {
    const nodeId = queue.shift() as string;
    ordered.push(byId.get(nodeId) as WorkflowNode);
    for (const target of downstream.get(nodeId) ?? []) {
      indegree.set(target, (indegree.get(target) ?? 0) - 1);
      if (indegree.get(target) === 0) queue.push(target);
    }
  }
  if (ordered.length !== nodes.length) throw new Error("Workflow must not contain cycles");
  return ordered;
}

export function edgeValue(edge: WorkflowEdge, values: Map<string, Record<string, unknown>>): unknown {
  const outputs = values.get(edge.source);
  if (!outputs) throw new Error(`Source node ${edge.source} has no outputs`);
  const port = edge.sourceHandle || "output";
  if (!(port in outputs)) throw new Error(`Source node ${edge.source} has no output port ${port}`);
  return outputs[port];
}

function upstreamValue(nodeId: string, workflow: Workflow, values: Map<string, Record<string, unknown>>): unknown {
  const incoming = dataEdges(workflow).filter((edge) => edge.target === nodeId);
  if (!incoming.length) return null;
  if (incoming.length > 1) throw new Error(`Node ${nodeId} currently accepts only one table input`);
  return edgeValue(incoming[0], values);
}

function upstreamTables(nodeId: string, workflow: Workflow, values: Map<string, Record<string, unknown>>): Record<string, Table> {
  const incoming = dataEdges(workflow).filter((edge) => edge.target === nodeId);
  const ports: Record<string, Table> = {};
  for (const edge of incoming) {
    const port = edge.targetHandle || "";
    if (port !== "left" && port !== "right") throw new Error(`Unknown concat input port: ${port}`);
    if (port in ports) throw new Error(`Concat input ${port} has more than one connection`);
    ports[port] = requireTable(edgeValue(edge, values), `Concat input ${port}`);
  }
  if (Object.keys(ports).length !== 2 || !("left" in ports) || !("right" in ports)) {
    throw new Error("Concat requires both A and B table inputs");
  }
  return ports;
}

function upstreamInputs(nodeId: string, workflow: Workflow, values: Map<string, Record<string, unknown>>): Record<string, unknown> {
  const incoming = dataEdges(workflow).filter((edge) => edge.target === nodeId);
  const inputs: Record<string, unknown> = {};
  for (const edge of incoming) {
    const port = edge.targetHandle || "input";
    if (port in inputs) throw new Error(`Input ${port} has more than one connection`);
    inputs[port] = edgeValue(edge, values);
  }
  return inputs;
}

export function nodeUpstream(nodeId: string, nodeType: string, workflow: Workflow, values: Map<string, Record<string, unknown>>): unknown {
  if (["table.concat", "table.merge_rows"].includes(nodeType)) return upstreamTables(nodeId, workflow, values);
  if (["table.concat_many", "pulse.combine_channels", "pulse.segment_measurement", "custom.python_function", "ui.alert", "function.call", "function.map", "logic.if_value", "logic.compare", "logic.switch", "math.operation", "logic.boolean_math"].includes(nodeType)) {
    return upstreamInputs(nodeId, workflow, values);
  }
  return upstreamValue(nodeId, workflow, values);
}

export function requireTable(value: unknown, operation: string): Table {
  if (!(value instanceof Table)) throw new Error(`${operation} requires a table input`);
  return value;
}

export function containerChildren(workflow: Workflow, containerId: string, branch: string): WorkflowNode[] {
  const children = workflow.nodes.filter((node) => node.parentId === containerId && (node.data.branch ?? "body") === branch);
  return children.sort((a, b) => {
    const ax = Number(a.position?.x ?? 0);
    const bx = Number(b.position?.x ?? 0);
    const ay = Number(a.position?.y ?? 0);
    const by = Number(b.position?.y ?? 0);
    return ax - bx || ay - by;
  });
}

export function flattenWorkflowGroups(workflow: Workflow): Workflow {
  const groups = new Map(workflow.nodes.filter((node) => node.data.nodeType === "workflow.group").map((node) => [node.id, node]));
  let flatEdges: WorkflowEdge[] = workflow.edges.map((edge) => ({ ...edge }));
  let changed = true;
  for (let pass = 0; changed && pass <= groups.size; pass += 1) {
    changed = false;
    const nextEdges: WorkflowEdge[] = [];
    for (const edge of flatEdges) {
      const targetGroup = groups.get(edge.target);
      if (targetGroup) {
        const port = (targetGroup.data.groupInputs ?? []).find((item) => item.id === edge.targetHandle);
        if (!port) throw new Error(`Group ${targetGroup.id} has no input port ${edge.targetHandle}`);
        nextEdges.push({ ...edge, target: String(port.internalNodeId), targetHandle: port.internalHandle ? String(port.internalHandle) : undefined });
        changed = true;
        continue;
      }
      const sourceGroup = groups.get(edge.source);
      if (sourceGroup) {
        const port = (sourceGroup.data.groupOutputs ?? []).find((item) => item.id === edge.sourceHandle);
        if (!port) throw new Error(`Group ${sourceGroup.id} has no output port ${edge.sourceHandle}`);
        nextEdges.push({ ...edge, source: String(port.internalNodeId), sourceHandle: port.internalHandle ? String(port.internalHandle) : undefined });
        changed = true;
        continue;
      }
      nextEdges.push(edge);
    }
    flatEdges = nextEdges;
  }
  if (changed) throw new Error("Workflow groups are nested too deeply or contain a port cycle");
  const flatNodes = workflow.nodes.filter((node) => node.data.nodeType !== "workflow.group").map((node) => {
    const data = { ...node.data };
    delete data.canvasParentId;
    return { ...node, data };
  });
  return { ...workflow, nodes: flatNodes, edges: flatEdges };
}

File: python/pydroid_flow/engine_parts/graph.py
from __future__ import annotations

from collections import defaultdict, deque
from typing import Any

import pandas as pd

from .values import _require_table

_NOTEBOOK_VISUAL_EDGE_ROLES = {"notebook-order", "notebook-variable", "notebook-parameter", "notebook-provenance"}

def _is_data_edge(edge: dict[str, Any]) -> bool:
    data = edge.get("data")
    role = data.get("role") if isinstance(data, dict) else None
    return role not in _NOTEBOOK_VISUAL_EDGE_ROLES

def _data_edges(workflow: dict[str, Any]) -> list[dict[str, Any]]:
    return [edge for edge in workflow.get("edges", []) if isinstance(edge, dict) and _is_data_edge(edge)]

def _ordered_nodes(workflow: dict[str, Any]) -> list[dict[str, Any]]:
    nodes = workflow.get("nodes", [])
    edges = _data_edges(workflow)
    if nodes and all("notebookCellIndex" in node.get("data", {}).get("parameters", {}) for node in nodes):
        return sorted(nodes, key=lambda node: (
            int(node.get("data", {}).get("parameters", {}).get("notebookCellIndex", 0)),
            int(node.get("data", {}).get("parameters", {}).get("notebookOperationIndex", 0)),
        ))
    by_id = {node["id"]: node for node in nodes}
    if len(by_id) != len(nodes):
        raise ValueError("Workflow node IDs must be unique")

    indegree = {node_id: 0 for node_id in by_id}
    downstream: dict[str, list[str]] = defaultdict(list)
    for edge in edges:
        source = edge["source"]
        target = edge["target"]
        if source not in by_id or target not in by_id:
            raise ValueError("Workflow contains an edge with a missing node")
        downstream[source].append(target)
        indegree[target] += 1

    queue = deque(node_id for node_id, count in indegree.items() if count == 0)
    ordered: list[dict[str, Any]] = []
    while queue:
        node_id = queue.popleft()
        ordered.append(by_id[node_id])
        for target in downstream[node_id]:
            indegree[target] -= 1
            if indegree[target] == 0:
                queue.append(target)

    if len(ordered) != len(nodes):
        raise ValueError("Workflow must not contain cycles")
    return ordered

def _edge_value(edge: dict[str, Any], values: dict[str, dict[str, Any]]) -> Any:
    outputs = values[edge["source"]]
    port = edge.get("sourceHandle") or "output"
    if port not in outputs:
        raise ValueError(f"Source node {edge['source']} has no output port {port}")
    return outputs[port]

def _upstream_value(node_id: str, workflow: dict[str, Any], values: dict[str, dict[str, Any]]) -> Any:
    incoming = [edge for edge in _data_edges(workflow) if edge["target"] == node_id]
    if not incoming:
        return None
    if len(incoming) > 1:
        raise ValueError(f"Node {node_id} currently accepts only one table input")
    return _edge_value(incoming[0], values)

def _upstream_tables(node_id: str, workflow: dict[str, Any], values: dict[str, dict[str, Any]]) -> dict[str, pd.DataFrame]:
    incoming = [edge for edge in _data_edges(workflow) if edge["target"] == node_id]
    ports: dict[str, pd.DataFrame] = {}
    for edge in incoming:
        port = edge.get("targetHandle") or ""
        if port not in {"left", "right"}:
            raise ValueError(f"Unknown concat input port: {port}")
        if port in ports:
            raise ValueError(f"Concat input {port} has more than one connection")
        ports[port] = _require_table(_edge_value(edge, values), f"Concat input {port}")
    if set(ports) != {"left", "right"}:
        raise ValueError("Concat requires both A and B table inputs")
    return ports

def _upstream_inputs(node_id: str, workflow: dict[str, Any], values: dict[str, dict[str, Any]]) -> dict[str, Any]:
    incoming = [edge for edge in _data_edges(workflow) if edge["target"] == node_id]
    inputs: dict[str, Any] = {}
    for edge in incoming:
        port = edge.get("targetHandle") or "input"
        if port in inputs:
            raise ValueError(f"Input {port} has more than one connection")
        inputs[port] = _edge_value(edge, values)
    return inputs

def _contained_node_ids(workflow: dict[str, Any]) -> set[str]:
    return {str(node["id"]) for node in workflow.get("nodes", []) if node.get("parentId")}

def _container_children(workflow: dict[str, Any], container_id: str, branch: str) -> list[dict[str, Any]]:
    children = [node for node in workflow.get("nodes", []) if node.get("parentId") == container_id and node.get("data", {}).get("branch", "body") == branch]
    return sorted(children, key=lambda node: (float(node.get("position", {}).get("x", 0)), float(node.get("position", {}).get("y", 0))))
